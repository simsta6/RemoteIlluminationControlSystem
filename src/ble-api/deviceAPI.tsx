import base64 from "base-64";
import { BleManager, Device as PlxDevice, Subscription } from "react-native-ble-plx";
import { BleDeviceActionTypes } from "../state/actions";
import { BleDevice } from "../state/ble-device/bleDeviceTypes";
import { Device } from "../state/devices/connectedDevicesTypes";
import { BLE_DEVICE_COMMANDS, generateMessage, ReadModuleParameter, TurnOnOrOffParameter } from "./messageGenerator";
import BluetoothStateManager from "react-native-bluetooth-state-manager";
import Toast from "react-native-toast-message";

type DevicesAdditionActionsType = {
    changeSvj: (newValue: number) => void;
    add: (device: Device) => void;
}

export class BleDeviceClient {
    private _bleManager: BleManager;
    private _bleDeviceId: string | undefined;
    private _actions: BleDeviceActionTypes;
    private _subscription: Subscription | undefined;
    private _serviceUUID: string | undefined;
    private _characteristicUUID: string | undefined;
    private _devicesAdditionActions: DevicesAdditionActionsType;
    private _tryingToConnect: boolean;

    public _didDeviceTriedToConnectOnStartup: boolean;

    constructor(bleManager: BleManager, bleDeviceId: string | undefined, serviceUUID: string, characteristicUUID: string, actions: BleDeviceActionTypes, devicesAdditionActions: DevicesAdditionActionsType) {
        this._bleManager = bleManager;
        this._bleDeviceId = bleDeviceId;
        this._actions = actions;
        this._devicesAdditionActions = devicesAdditionActions;
        this._subscription = undefined;
        this._serviceUUID = serviceUUID;
        this._characteristicUUID = characteristicUUID;

        this._tryingToConnect = false;
        this._didDeviceTriedToConnectOnStartup = false;
        bleDeviceId && this.connectToDevice(bleDeviceId).then(() => {
            this._didDeviceTriedToConnectOnStartup = true;
        });
    }

    public readResponse(command: string, bleDevice: BleDevice) {
        const response = bleDevice.messages.find(msg => msg.command.includes(command))?.response;
        this._actions.readResponse(command);
        return response;
    }

    public async connectToDevice(deviceId: string): Promise<boolean> {
        let connectedDevice: PlxDevice | undefined = undefined;
        const bleState = await BluetoothStateManager.getState();
        if (bleState === "PoweredOff") {
            await  BluetoothStateManager.requestToEnable();
        }

        if (!await this._bleManager.isDeviceConnected(deviceId) && !this._tryingToConnect) {
            this._tryingToConnect = true;
            connectedDevice = await this._bleManager
                .connectToDevice(deviceId)
                .then(cDevice => {
                    return cDevice.discoverAllServicesAndCharacteristics().then(async (deviceWithChar) => {
                        console.log("connected");
                        this._bleDeviceId = deviceWithChar.id;
                        await this.setPrimaryCharacteristic(this._bleManager, deviceWithChar.id);
                        await this.listenToMessages(deviceWithChar.id);
                        this._tryingToConnect = false;
                        return deviceWithChar;
                    });
                })
                .catch(err => {
                    this.showError(err);
                    return undefined;
                });
        }

        connectedDevice && this._actions.modify({
            deviceId: connectedDevice.id,
            isDeviceConnected: !!(connectedDevice.id && this._subscription),
            messages: [],
            serviceUUID: this._serviceUUID,
            uuid: this._characteristicUUID,
        });

        return connectedDevice ? true : false;
    }

    public async sendMessage(message: string, needResponse = true, timesToTry = 2): Promise<boolean> {
        const characteristicsUUID = this._characteristicUUID;
        const serviceUUID = this._serviceUUID;
        const bleDeviceId = this._bleDeviceId;
        if (!characteristicsUUID || !serviceUUID || !bleDeviceId) return false;

        if (await this._bleManager.isDeviceConnected(bleDeviceId) && this._subscription) {
            await this._bleManager.writeCharacteristicWithResponseForDevice(bleDeviceId, serviceUUID, characteristicsUUID, base64.encode(message + "\r\n"))
                .then((char) => {
                    if (char.value) {
                        const command = base64.decode(char.value);
                        needResponse && this._actions.sendRequest(command);
                    }
                })
                .catch(err => {
                    this.showError(err);
                    return false;
                });
            return true; // to prevent sending message if this client did not tried to connect to device 
        } else {
            if (timesToTry < 1) return false;
            await this.connectToDevice(bleDeviceId);
            const res = await this.sendMessage(message, needResponse, timesToTry - 1);
            return res;
        }
    }

    public get bleManager() {
        return this._bleManager;
    }

    public get didDeviceTriedToConnectOnStartup() {
        return this._didDeviceTriedToConnectOnStartup;
    }

    public get bleDeviceId() {
        return this._bleDeviceId;
    }

    public async requestStats(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.GetStats);
        return await this.sendMessage(message, false);
    }

    public async sendMessageToGetModuleColor(moduleId: string): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.ReadModule, moduleId, ReadModuleParameter.GetRgbColor);
        const isSuccessful = await this.sendMessage(message);
        return isSuccessful;
    }

    public async changeDeviceColorOrBrightness(moduleIdOrIds: string[], colors: { [x: string]: string; }): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const color = colors[id];
            const message = generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, id, color.substring(1, 7));
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOnSpecificDevice(moduleIdOrIds: string[]): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnOrOffModule, id, TurnOnOrOffParameter.ON);
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOffSpecificDevice(moduleIdOrIds: string[]): Promise<boolean> {
        const result = await Promise.all(moduleIdOrIds.map(async id => {
            const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnOrOffModule, id, TurnOnOrOffParameter.OFF);
            return await this.sendMessage(message, false);
        }));
        return !result.some(res => !res);
    }

    public async turnOnAllDevices(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOnAllModules);
        return await this.sendMessage(message, false);
    }

    public async turnOffAllDevices(): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.TurnOffAllModules);
        return await this.sendMessage(message, false);
    }

    public async testDeviceByBlinking(moduleId: string, initColor: string): Promise<boolean> {
        await this.sendMessage(generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, moduleId, "050505"), false);
        await new Promise(r => setTimeout(r, 500));
        const result =  await this.sendMessage(generateMessage(BLE_DEVICE_COMMANDS.ChangeColorOrBrightness, moduleId, initColor.substring(1, 7)), false);
        return result;
    }

    public async changeLightSensorValue(newValue: number): Promise<boolean> {
        const message = generateMessage(BLE_DEVICE_COMMANDS.ChangeLightSensorValue, undefined, Math.round(newValue).toString(16));
        return await this.sendMessage(message, false);
    }

    private async listenToMessages(deviceId: string) {
        this._subscription?.remove();
        const characteristicsUUID = this._characteristicUUID;
        const serviceUUID = this._serviceUUID;

        if (characteristicsUUID && serviceUUID) {
            this._subscription = await this._bleManager.isDeviceConnected(deviceId)
                .then(isConnected => {
                    if (isConnected) {
                        return this._bleManager.monitorCharacteristicForDevice(deviceId, serviceUUID, characteristicsUUID, (error, char) => {
                            if (error) {
                                if (`${error.name}: ${error.message}`.includes("BleError: Device") && error.message.includes("was disconnected")
                                    || `${error.name}: ${error.message}`.includes("BleError: Characteristic") && error.message.includes("notify change failed for device")) {
                                    return;
                                }

                                this.showError(error.message);
                                return;
                            }
                            if (char?.value) {
                                const response = base64.decode(char.value).replace(/(\r\n|\n|\r)/gm,"");
                                const responseArray = response.split(":");
                                if (responseArray.length > 1) {
                                    const command = responseArray[0];
                                    const response = responseArray.slice(1);
                                    if (command.includes(BLE_DEVICE_COMMANDS.GetStats)) {
                                        const values = response.join(",").split(",");
                                        if (values[0].includes("SVJ")) {
                                            this._devicesAdditionActions.changeSvj(parseInt(values[2], 16));
                                        } else {
                                            const index = values[0].slice(2);
                                            const color = "#" + values[1].toUpperCase();
                                            const current = values[5] === "0" ? 0.045 : 0.0045;
                                            const voltage = parseInt(values[3], 16) / 1000;
                                            const power = current * voltage;
                                            const temperature = parseInt(values[4], 16).toString();
                                            const bulbType = values[5] === "0" ? "Non-RGB" : "RGB";
                                            
                                            this._devicesAdditionActions.add({
                                                index,
                                                color,
                                                bulbType,
                                                temperature,
                                                power: power.toString(),
                                                voltage: voltage.toString(),
                                                current: current.toString(),
                                            });
                                        }
                                    } else {
                                        this._actions.updateResponse(command, response);
                                    }
                                }

                                console.log(response);

                            }
                        });
                    }
                })
                .catch(err => {
                    this.showError(err);
                    return undefined;
                });
        }
    }

    private async setPrimaryCharacteristic(bleManager: BleManager, deviceId: string): Promise<boolean> {
        const services = await bleManager.servicesForDevice(deviceId);

        for (const service of services) {
            const characteristics = await bleManager.characteristicsForDevice(deviceId, service.uuid);

            for (const characteristic of characteristics) {
                if (characteristic.isWritableWithResponse) {
                    this._serviceUUID = characteristic.serviceUUID;
                    this._characteristicUUID = characteristic.uuid;
                    return true;
                }
            }
        }
        return false;
    }

    public async sendMessageWithValidation(message: string): Promise<boolean> {
        if (this.validateCommand(message)) {
            return this.sendMessage(message, false);
        }
        return Promise.resolve(false);
    }

    private validateCommand(command: string): boolean {
        const regexExpressions = [
            "LIST",
            "STATS",
            "ON",
            "OFF",
            "OK",
            "ID[0-9A-F]{1,2}CLR[0-9A-F]{6}",
            "ID[0-9A-F]{1,2}ST[1-5]",
            "ID[0-9A-F]{1,2}ONF[0-1]",
            "LX[0-9A-F]{1,3}"
        ] as const;

        for(let i = 0; i < regexExpressions.length; i++) {
            const re = new RegExp(regexExpressions[i]);
            if (re.exec(command)) {
                return true;
            }
        }

        return false;
    }

    private showError(err: string) {
        Toast.show({
            type: "error",
            text1: "Error",
            text2: err
        });
    }
}