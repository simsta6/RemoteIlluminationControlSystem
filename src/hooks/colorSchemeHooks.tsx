import { useTheme } from "@react-navigation/native";
import { AppDarkColors, AppLightColors } from "../constants/themes";

export const useAppColors = () => {
    const { dark } = useTheme();

    return {
        colors: dark ? AppDarkColors : AppLightColors
    };
};
