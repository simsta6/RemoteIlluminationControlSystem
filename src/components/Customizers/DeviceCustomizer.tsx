import { Slider } from "@miblanchard/react-native-slider";
import React from "react";
import { StyleSheet, View } from "react-native";
import SunIcon from "../../assets/icons/SunIcon";
import { BleDeviceClient } from "../../ble-api/deviceAPI";
import { isHexColor, shadeColorIfNeeded } from "../../helpers/colorHelper";
import { DevicesKeys, getDevicesIdBySelectedDevice } from "../../helpers/devicesHelper";
import { useBleDevice } from "../../hooks/bleDeviceHook";
import { useAppColors } from "../../hooks/colorSchemeHooks";
import { useConnectedDevices } from "../../hooks/connectedDevicesHooks";
import { Device } from "../../state/devices/connectedDevicesTypes";
import { Button } from "../Buttons/Button";
import { ColorPicker, ColorPickerProps } from "../ColorPicker";

const isColorPickerNeeded = (devices: Device[], selectedDevice: string): boolean => { 
    switch (selectedDevice) {
    case DevicesKeys.RgbDevices:
    case DevicesKeys.AllDevices:
    case DevicesKeys.NonRgbDevices: 
        return false;
    default:
        return devices.find(dev => dev.index === selectedDevice)?.bulbType === "RGB";
    }
};

const getColorPickerIfNeeded = (devices: Device[], selectedDevice: string, colorPickerProps: ColorPickerProps) => {
    console.log("SelectedDevice: " + selectedDevice);
    console.log("Color: " + colorPickerProps.color);
    return isColorPickerNeeded(devices, selectedDevice) ? (
        <View style={{marginVertical: 10, width: "100%"}}>
            <ColorPicker {...colorPickerProps} />
        </View>
    ) : <></>;
};

interface Props {
    setIsConnectDevicesModalVisible: React.Dispatch<React.SetStateAction<boolean>>;
    bleDeviceClient: BleDeviceClient;
    selectedDevice: string;
}

export const DeviceCustomizer = (props: Props) => {
    const { setIsConnectDevicesModalVisible, bleDeviceClient, selectedDevice } = props;
    const { colors: themeColors } = useAppColors();

    const [devices, devicesActions] = useConnectedDevices();
    const [bleDevice] = useBleDevice();

    const [color, setColor] = React.useState(devices.find(dev => dev.index === selectedDevice)?.color ?? "#FFFFFF");
    const [sliderValue, setSliderValue] = React.useState(100);
    const [isRGBDevice, setIsRGBDevice] = React.useState(false);

    const bleDeviceId = bleDevice.deviceId;

    React.useEffect(() => {
        setIsRGBDevice(isColorPickerNeeded(devices, selectedDevice));
    }, [selectedDevice]);

    React.useEffect(() => {
        const newColor = isRGBDevice && devices.find(dev => dev.index === selectedDevice)?.color;
        newColor && isHexColor(newColor) && setColor(newColor);
    }, [isRGBDevice, selectedDevice]);

    React.useEffect(() => {
        let setHook = true;
        const timeOut = setTimeout(async () => {
            if (bleDeviceId) {
                const ids = getDevicesIdBySelectedDevice(devices, selectedDevice);
                const colors = isRGBDevice ? [{[selectedDevice]: shadeColorIfNeeded(color, sliderValue)}] :
                    ids.map(id => ({ [id]: shadeColorIfNeeded(devices.find(dev => dev.index === id)?.color ?? "#FFFFFF", sliderValue) }));

                const messageSent = await bleDeviceClient.changeDeviceColorOrBrightness(ids, Object.assign({}, ...colors ));
                messageSent && devicesActions.changeColor(selectedDevice, color);
                !messageSent && setHook && setIsConnectDevicesModalVisible(true);
            } else {
                setHook && setIsConnectDevicesModalVisible(true);
            }
        }, 200);
        return () => {
            clearTimeout(timeOut);
            setHook = false;
        };
    }, [color, sliderValue]);

    return (
        <>
            <Slider
                containerStyle={{width: "100%", marginTop: 6}}
                minimumValue={0}
                maximumValue={100}
                value={ sliderValue }
                trackStyle={ styles.trackStyle }
                minimumTrackTintColor={ themeColors.text }
                maximumTrackTintColor={ themeColors.card }
                renderThumbComponent={() => 
                    <View style={{...styles.thumbContainer, backgroundColor: themeColors.background, borderColor: themeColors.border}}>
                        <SunIcon color={themeColors.icon} height={30} width={30}/>
                    </View>
                }
                thumbTouchSize={{
                    width: 50,
                    height: 50,
                }}
                onValueChange={v => {
                    const value = v instanceof Array ? v[0] : v;
                    setSliderValue(value);
                }}
            />
            <Button title="Test" onPress={() => console.log(bleDeviceClient.requestStats())}></Button>
            {getColorPickerIfNeeded(devices, selectedDevice, {...{color, setColor}})}
            <Button title="show messages" onPress={() => bleDevice.messages.forEach(a=>console.log(a))} />
            <Button title="remove devices" onPress={() => {
                devices.forEach((_, index) => devicesActions.remove(index));
            }} />
        </>
    );
};

const styles = StyleSheet.create({
    thumbContainer: {
        height: 35,
        width: 35,
        borderRadius: 100,
        borderWidth: 1,
        alignItems: "center",
        justifyContent: "center"
    },
    trackStyle: { 
        height: 10, 
        borderRadius: 10,
    },
});