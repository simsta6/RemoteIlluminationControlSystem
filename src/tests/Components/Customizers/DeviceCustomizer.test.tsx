import "@testing-library/jest-dom";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { BleDeviceClient } from "../../../ble-api/deviceAPI";
import { DeviceCustomizer } from "../../../components/Customizers/DeviceCustomizer";
import { Colors } from "../../../constants/themes";
import * as bleDeviceHook from "../../../hooks/bleDeviceHook";
import * as colorSchemeHooks from "../../../hooks/colorSchemeHooks";
import * as connectedDevicesHooks from "../../../hooks/connectedDevicesHooks";
import { BleDevice } from "../../../state/ble-device/bleDeviceTypes";
import { Device } from "../../../state/devices/connectedDevicesTypes";

describe("<DeviceCustomizer />", () => {
    const colors = {
        icon: "orange",
        text: "purple",
        card: "white",
        background: "red",
        border: "black",
    };

    const devices: Device[] = [{
        index: "1",
        color: "#ffff00",
        bulbType: "RGB",
        temperature: "16",
        power: "15",
        voltage: "3",
        current: "5",
        name: "RGB Device" 
    },
    {
        index: "2",
        color: "#ffff00",
        bulbType: "Non-RGB",
        temperature: "16",
        power: "15",
        voltage: "3",
        current: "5",
        name: "Non-RGB Device" 
    }];

    const bleDevice: BleDevice = {
        deviceId: "bleDeviceId",
        messages: [],
        isDeviceConnected: true,
        serviceUUID: "",
        subscription: undefined,
        uuid: "",
    };

    let mockChangeDeviceColorOrBrightness: jest.SpyInstance<Promise<boolean>, [moduleIdOrIds: string[], colors: {
        [x: string]: string;
    }]>;

    let mockSetIsConnectDevicesModalVisible: jest.Mock;

    const mockedBleDeviceClient = BleDeviceClient.prototype;

    beforeEach(() => {
        jest.setTimeout(100 * 1000);
        mockChangeDeviceColorOrBrightness = jest
            .spyOn(BleDeviceClient.prototype, "changeDeviceColorOrBrightness")
            .mockImplementation(() => Promise.resolve(true));

        jest.spyOn(BleDeviceClient.prototype, "didDeviceTriedToConnectOnStartup", "get").mockImplementation( () => true);

        mockSetIsConnectDevicesModalVisible = jest.fn().mockImplementation();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    beforeAll(() => {
        jest.spyOn(colorSchemeHooks, "useAppColors").mockImplementation(() => ({ colors: colors as Colors }));

        jest.spyOn(connectedDevicesHooks, "useConnectedDevices").mockImplementation(() => [devices, {
            add: () => ({}),
            remove: () => ({}),
            removeAll: () => ({}),
            changeColor: (deviceIndex: string, color: string) => {
                devices.map(dev => dev.index === deviceIndex ? {...dev, color } : dev);
            },
            modify: () => ({}),
        }]);

        jest.spyOn(bleDeviceHook, "useBleDevice").mockImplementation(() => [bleDevice, {
            remove: () => ({}),
            sendRequest: () => ({}),
            readResponse: () => ({}),
            updateResponse: () => ({}),
            modify: () => ({}),
        }]);
    });

    test("Default DeviceCustomizer should match snapshot", () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: BleDeviceClient.prototype,
            selectedDevice: "1",
        };
        const { toJSON } = render(
            <DeviceCustomizer {...props} />
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test("should call changeDeviceColorOrBrightness twice when color was changed ", async () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: mockedBleDeviceClient,
            selectedDevice: "1",
        };
        const root = await waitFor(() => { 
            return render(<DeviceCustomizer {...props} />);
        });

        const redSwatch = root.getByTestId("#ff0000");
        fireEvent.press(redSwatch);

        await waitFor(() => {
            expect(mockChangeDeviceColorOrBrightness).toHaveBeenLastCalledWith(["1"], {"1": "#ff0000"});
            expect(mockChangeDeviceColorOrBrightness).toHaveBeenCalledTimes(2);
        });   
    });

    test("swatches should be missing if selected device is Non-RGB", async () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: mockedBleDeviceClient,
            selectedDevice: "2",
        };
        const root = await waitFor(() => { 
            return render(<DeviceCustomizer {...props} />);
        });

        await waitFor(() => {
            expect(root.queryByTestId("#ff0000")).not.toBeInTheDocument();
        });
    });

    test("swatches should be missing when group of devices is selected", async () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: mockedBleDeviceClient,
            selectedDevice: "AllDevices",
        };
        const root = await waitFor(() => { 
            return render(<DeviceCustomizer {...props} />);
        });

        await waitFor(() => {
            expect(root.queryByTestId("#ff0000")).not.toBeInTheDocument();
        });
    });

    test("should call changeDeviceColorOrBrightness 4 times when group of devices is selected", async () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: mockedBleDeviceClient,
            selectedDevice: "AllDevices",
        };

        await waitFor(() => { 
            return render(<DeviceCustomizer {...props} />);
        });

        await waitFor(() => {
            expect(mockChangeDeviceColorOrBrightness).toHaveBeenCalledTimes(4);
        });   
    });

    test("should call changeDeviceColorOrBrightness with multiple ids when group of ddevices is selected", async () => {
        const props = {
            setIsConnectDevicesModalVisible: mockSetIsConnectDevicesModalVisible as () => (unknown) as React.Dispatch<React.SetStateAction<boolean>>,
            bleDeviceClient: mockedBleDeviceClient,
            selectedDevice: "AllDevices",
        };

        await waitFor(() => { 
            return render(<DeviceCustomizer {...props} />);
        });

        await waitFor(() => {
            expect(mockChangeDeviceColorOrBrightness).toHaveBeenLastCalledWith(["1", "2"], {"1": "#ffff00", "2": "#ffff00"});
        });    
    });

});