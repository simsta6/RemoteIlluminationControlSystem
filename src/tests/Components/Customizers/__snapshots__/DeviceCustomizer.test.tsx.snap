// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceCustomizer /> Default DeviceCustomizer should match snapshot 1`] = `
Array [
  <View
    animationType="timing"
    onLayout={[Function]}
    onValueChange={[Function]}
    step={0}
    style={
      Array [
        Object {
          "height": 40,
          "justifyContent": "center",
        },
        Object {},
        Object {
          "marginTop": 6,
          "width": "100%",
        },
      ]
    }
    thumbTouchSize={
      Object {
        "height": 50,
        "width": 50,
      }
    }
    trackClickable={true}
    trackMarks={Array []}
    value={100}
  >
    <View
      onLayout={[Function]}
      renderToHardwareTextureAndroid={true}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "height": 4,
          },
          Object {
            "backgroundColor": "white",
          },
          Object {
            "borderRadius": 10,
            "height": 10,
          },
        ]
      }
    />
    <View
      collapsable={false}
      renderToHardwareTextureAndroid={true}
      style={
        Object {
          "backgroundColor": "purple",
          "borderRadius": 10,
          "height": 10,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      collapsable={false}
      onLayout={[Function]}
      style={
        Object {
          "opacity": 0,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "red",
            "borderColor": "black",
            "borderRadius": 100,
            "borderWidth": 1,
            "height": 35,
            "justifyContent": "center",
            "width": 35,
          }
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={30}
          bbWidth={30}
          focusable={false}
          height={30}
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              Object {
                "flex": 0,
                "height": 30,
                "width": 30,
              },
            ]
          }
          testID="SunIcon"
          vbHeight={512}
          vbWidth={512}
          width={30}
          x="0"
          y="0"
        >
          <RNSVGGroup>
            <RNSVGPath
              d="M256 105.5c-83.9 0-152.2 68.3-152.2 152.2 0 83.9 68.3 152.2 152.2 152.2 83.9 0 152.2-68.3 152.2-152.2 0-84-68.3-152.2-152.2-152.2zm0 263.5c-61.4 0-111.4-50-111.4-111.4 0-61.4 50-111.4 111.4-111.4 61.4 0 111.4 50 111.4 111.4 0 61.4-50 111.4-111.4 111.4zM256 74.8c11.3 0 20.4-9.1 20.4-20.4v-23c0-11.3-9.1-20.4-20.4-20.4-11.3 0-20.4 9.1-20.4 20.4v23c0 11.3 9.1 20.4 20.4 20.4zM256 437.2c-11.3 0-20.4 9.1-20.4 20.4v22.9c0 11.3 9.1 20.4 20.4 20.4 11.3 0 20.4-9.1 20.4-20.4v-22.9c0-11.2-9.1-20.4-20.4-20.4zM480.6 235.6h-23c-11.3 0-20.4 9.1-20.4 20.4 0 11.3 9.1 20.4 20.4 20.4h23c11.3 0 20.4-9.1 20.4-20.4 0-11.3-9.1-20.4-20.4-20.4zM54.4 235.6h-23c-11.3 0-20.4 9.1-20.4 20.4 0 11.3 9.1 20.4 20.4 20.4h22.9c11.3 0 20.4-9.1 20.4-20.4.1-11.3-9.1-20.4-20.3-20.4zM400.4 82.8L384.1 99c-8 8-8 20.9 0 28.9s20.9 8 28.9 0l16.2-16.2c8-8 8-20.9 0-28.9s-20.9-8-28.8 0zM99 384.1l-16.2 16.2c-8 8-8 20.9 0 28.9s20.9 8 28.9 0l16.2-16.2c8-8 8-20.9 0-28.9s-20.9-7.9-28.9 0zM413 384.1c-8-8-20.9-8-28.9 0s-8 20.9 0 28.9l16.2 16.2c8 8 20.9 8 28.9 0s8-20.9 0-28.9L413 384.1zM99 127.9c8 8 20.9 8 28.9 0s8-20.9 0-28.9l-16.2-16.2c-8-8-20.9-8-28.9 0s-8 20.9 0 28.9L99 127.9z"
              fill={4294944000}
              propList={
                Array [
                  "fill",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "marginBottom": -20,
            "marginLeft": -20,
            "marginRight": -20,
            "marginTop": -20,
          },
        ]
      }
    />
  </View>,
  <View
    style={
      Object {
        "marginVertical": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 10,
        }
      }
    >
      <View
        style={
          Object {
            "height": 718,
            "width": 718,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "space-between",
                "overflow": "visible",
              },
              Object {},
              undefined,
            ]
          }
        >
          <View
            onLayout={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "minHeight": 200,
                  "minWidth": 200,
                  "overflow": "visible",
                  "position": "relative",
                  "width": "100%",
                },
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 15,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "marginTop": 10,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#ff0000"
        >
          <View
            style={
              Object {
                "backgroundColor": "#ff0000",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#1633e6"
        >
          <View
            style={
              Object {
                "backgroundColor": "#1633e6",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#00c85d"
        >
          <View
            style={
              Object {
                "backgroundColor": "#00c85d",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#ffde17"
        >
          <View
            style={
              Object {
                "backgroundColor": "#ffde17",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#98b048"
        >
          <View
            style={
              Object {
                "backgroundColor": "#98b048",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="#3a9cb0"
        >
          <View
            style={
              Object {
                "backgroundColor": "#3a9cb0",
                "borderRadius": 6,
                "height": 30,
                "width": 30,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>,
]
`;
